### PR TITLE
Eval Wrapper Refactor

### PR DESCRIPTION
**Pull request by julianrocha**
Refactor to help make the evaluation of many models in Google Collab easier. Contains changes from both Robert and I

## New features
In addition to the changes mentioned below, new features include
- run model on a folder of images or a single image
- average predictions made by running the original image and a left/right flipped image through the network to improve results
- fix a number of memory leaks (but still not entirely fixed)
- fix stacked heatmap ground truth having a white border 

# Principles behind this refactor
Below is a summary of the principles we tried to follow for this refactor.

## Rules
- predict takes X_batch
- other methods take predicted_heatmaps_batch
- OKS/PCK take full_list_of_predictions
- limit methods within eval calling other public methods

## Notebook to run in Google Collab (`model_tester.ipynb`)
- Run in collab
- specify eval action
- keep simple and flexible
- be able to show plots/visualizations

## Wrapper (`evaluation_wrapper.py`)
- iterating over epochs of model (making new instance)
- preparing data to evaluate (representative set OR folder of images)
- creates data generator and saves it internally (only need to create once)
- some form of cached results/predictions in memory

## Evaluation Public methods(on single model) (`evaluation.py`)
- visualize (heatmaps stacked) predicted_heatmaps_batch -> pngs
- visualize (scatterplot on GT scale image, along with ground truth) predicted_heatmaps_batch -> pngs
- return predicted heatmaps for a single batch
- return predictions in COCO format: predicted_heatmaps_batch -> list of dicts (need to append lists in wrapper)
- perform OKS (): full_list_of_predictions -> append to csv, with epoch, for model
- perform PCK (): full_list_of_predictions ->  append to csv, with epoch, for model
- plotPCK/OKS performance over epochs: csv path -> matplot

### CODE REVIEW

**Comment by robertklee** [comment]
Looks great so far! I'll merge with your changes once I'm ready.

File: model_tester.ipynb [hunk 1]
```diff
@@ -0,0 +1 @@
...AB_TRAINING variable in constants file\n","# https://askubuntu.com/questions/20414/find-and-replace-text-within-a-file-using-commands\n","!sed -i.bak 's/COLAB_TRAINING = False/COLAB_TRAINING = True/g' constants.py \n","\n","# Remove backup file\n","!rm constants.py.bak\n","\n","# Upgrade required imgaug package because of a bug on 0.2.9\n","!pip install --upgrade imgaug==0.4.0"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"CurWEGcdRHtA"},"source":["import imp\n","import evaluation_wrapper\n","#imp.reload(evaluation_wrapper)\n","import evaluation\n","#imp.reload(evaluation)\n","\n","eval = evaluation_wrapper.EvaluationWrapper('2021-04-02-02h-21m_batchsize_16_hg_4_loss_keras_mse_aug_medium_sigma4_learningrate_5.0e-03_opt_adam_gt-4kp_activ_linear')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"LZe2owx2RHtG"},"source":["eval.calculateOKS(1, 'val_set')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"4CnUMzg8RHtH"},"source":["eval.calculatePCK(1, 'val_set')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"GeQHWFyMRHtH"},"source":[""],"execution_count":null,"outputs":[]}]}
```
Comment (robertklee) [inline model_tester.ipynb]: <p>I'm wondering if we should just add these cells to the Colab train notebook. We can choose to not run the training cell.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/robertklee/COCO-Human-Pose/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (julianrocha) [reply model_tester.ipynb]: good call

File: colab_train.ipynb [hunk 2]
```diff
@@ -8,8 +8,13 @@
       "collapsed_sections": []
     },
     "kernelspec": {
-      "display_name": "Python 3",
-      "name": "python3"
+      "name": "python3",
+      "display_name": "Python 3.8.2 64-bit",
```
Comment (robertklee) [inline colab_train.ipynb]: we might want to revert this change since we have been using python 3.7 in colab

File: colab_train.ipynb [hunk 3]
```diff
@@ -240,6 +245,47 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "# Run OKS\n",
+        "\n",
+        "eval.calculateOKS(1, 'val_set')"
```
Comment (robertklee) [inline colab_train.ipynb]: btw I changed to an Enum type. `Generator.val_gen` from `constants`

File: evaluation_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        return pck
+
+    def plotOKS(self, epochs):
+        pass
+
+    def plotPCK(self, epochs):
+        pass
+
+    @lru_cache(maxsize=50)
+    def _full_list_of_predictions(self, gen, model_sub_dir, epoch):
```
Comment (robertklee) [inline evaluation_wrapper.py]: is this supposed to take `model_sub_dir, epoch`?

File: colab_train.ipynb (hunk 2 above)
Comment (robertklee) [reply colab_train.ipynb]: Reverted. Closing this comment now

File: evaluation_wrapper.py (hunk 4 above)
Comment (julianrocha) [reply evaluation_wrapper.py]: It does this to make calls to it unique, this way the cache can be used

Comment (robertklee) [reply evaluation_wrapper.py]: ahh makes sense.

File: run_evaluation.py [hunk 5]
```diff
@@ -9,21 +9,21 @@
 # %matplotlib inline
```
Comment (robertklee) [inline run_evaluation.py]: should we move these `run_evaluation` and `run_visualization` into an outdated folder?

**Review by robertklee — APPROVED** [review]
LGTM! I think this is a great refactor!

File: run_evaluation.py (hunk 5 above)
Comment (julianrocha) [reply run_evaluation.py]: yes I'll do that now
